pkg/compact/dedup.go

```diff
@@ -0,0 +1,322 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package compact
+
+import (
+	"container/heap"
+	"math"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+)
+
+// NewDedupChunkSeriesMerger merges several chunk series into one.
```

Comment:
I assume you copied some of things over to next file? Can you mark which of those functions/methods are unchanged?

Reply:
I assume nothing has changed here

Reply:
This is mainly a cleanup and I am also wondering is there a better way to do it.
1. The original files only works for SeriesSet and the compact merger function requires ChunkSeriesSet. So I cannot directly use the original functions.
2. The original functions have logic for counter reset and adjustment. This is not needed in the compactor side.
3. I need a place to put some code for the encoder and decoder of chunk and series. Do you think it reasonable if I put them in the `query` package?

Reply:
Well we already have some code for there, but you are right, it's something both compaction and querier would use. Maybe shared dedup package would do?

Reply:
Yes, a shared package makes sense.

Reply:
As this change is mainly a cleanup, I split it to a new pr https://github.com/thanos-io/thanos/pull/4242.